### PR TITLE
Optimize pending list processing

### PR DESCRIPTION
**Pull request by wez**
This is a series of diffs to improve our handling of pending list processing.

There are a couple of diffs that are simple refactorings to libart, but there are two that bear special scrutiny:

* `libart: better defined behavior when a full prefix key is encountered`
* `use art to optimize pending list construction`

The former is necessary to avoid being forced to make a copy and NUL-terminate all input strings for both lookup, prefix iteration and insertion.

The latter is really the main component of this pull request.

In the included benchmark we can reduce a bloated pending list full of 260k pending lstat operations down to a dozen or so recursive walks.  This is advantageous because it eliminates a large number of redundant lstat calls and should give us lower latency and higher throughput when processing notifications from the kernel.

That benchmark changes the 2 seconds it takes for 260k lstat calls for todays implementation to a fraction of a second with this more optimal version.

In addition to the individual test plans in these diffs, I run through the mercurial test suite with the tip of this branch:

```
 WATCHMAN_SOCK=/opt/facebook/watchman/var/run/watchman/wez-state/sock ./run-tests.py --blacklist=blacklists/fsmonitor --extra-config="extensions.fsmonitor=" -j 3
```

and observed no new failures vs. the currently deployed watchman.

### CODE REVIEW

**Comment by facebook-github-bot-8** [comment]
By analyzing the blame information on this pull request, we identified @Vaelatern and @anirbanr-fb to be potential reviewers.

**Comment by facebook-github-bot-5** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-1** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@wez updated the pull request.

File: thirdparty/libart/src/art.c [hunk 1]
```diff
@@ -21,6 +33,44 @@ static uint32_t __inline __builtin_ctz(uint32_t x) {
 }
 #endif
 
+// The ART implementation requires that no key be a full prefix of an existing
+// tree during insertion.  In practice this means that each key must have a
```
Comment (sid0) [inline thirdparty/libart/src/art.c]: `s/tree/key`

File: thirdparty/libart/src/art.c [hunk 2]
```diff
@@ -215,15 +265,15 @@ static int check_prefix(const art_node *n, const unsigned char *key, int key_len
 
 /**
  * Checks if a leaf matches
- * @return 0 on success.
+ * @return true if the key is an exact match.
  */
-static int leaf_matches(const art_leaf *n, const unsigned char *key, int key_len, int depth) {
-    (void)depth;
+static bool leaf_matches(const art_leaf *n, const unsigned char *key, int key_len) {
```
Comment (sid0) [inline thirdparty/libart/src/art.c]: We've effectively forked libart at this point, then :(

File: thirdparty/libart/src/art.c [hunk 3]
```diff
@@ -21,6 +33,44 @@ static uint32_t __inline __builtin_ctz(uint32_t x) {
 }
 #endif
 
+// The ART implementation requires that no key be a full prefix of an existing
+// tree during insertion.  In practice this means that each key must have a
+// terminator character.  One approach is to ensure that the key and key_len
+// includes a physical trailing NUL terminator when inserting C-strings.
+// This doesn't help a great deal when working with binary strings that may be
+// a slice in the middle of a buffer that has no termination.
```
Comment (sid0) [inline thirdparty/libart/src/art.c]: Ah, so you basically want to avoid making a copy of each binary string then?

File: thirdparty/libart/src/art.c [hunk 4]
```diff
@@ -21,6 +33,44 @@ static uint32_t __inline __builtin_ctz(uint32_t x) {
...
+//
+// To facilitate this the key_at() function is used to look up the byte
+// value at a given index.  If that index is 1 byte after the end of the
+// key, we synthesize a fake NUL terminator byte.
```
Comment (sid0) [inline thirdparty/libart/src/art.c]: I'd add a note here that this still assumes that the binary strings do not have embedded null bytes (which is valid because we only insert paths, and paths can never contain null bytes). If the binary strings did have embedded NULs then this approach wouldn't work.

File: thirdparty/libart/src/art.c [hunk 5]
```diff
@@ -21,6 +33,44 @@ static uint32_t __inline __builtin_ctz(uint32_t x) {
...
+// @return the value of the key at the supplied index.
+static inline unsigned char key_at(const unsigned char *key, int key_len, int idx) {
+    if (idx == key_len) {
+        // Implicit terminator
+        return 0;
+    }
+#if !ART_SANITIZE_ADDRESS
+    // If we were built with -fsanitize=address, let ASAN catch this,
+    // otherwise, make sure we blow up if the input depth is out of bounds.
+    w_assert(idx >= 0 && idx <= key_len,
```
Comment (sid0) [inline thirdparty/libart/src/art.c]: Does this assert work in release builds?

File: thirdparty/libart/src/art.c [hunk 6]
```diff
@@ -241,10 +291,10 @@ void* art_search(const art_tree *t, const unsigned char *key, int key_len) {
     while (n) {
         // Might be a leaf
         if (IS_LEAF(n)) {
-            n = (art_node*)LEAF_RAW(n);
+            art_leaf *leaf = LEAF_RAW(n);
```
Comment (sid0) [inline thirdparty/libart/src/art.c]: Are these changes related to this diff? Having these included is making this harder to read.

File: watchman_log.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+void w_log(int level, WATCHMAN_FMT_STRING(const char *fmt), ...)
+#ifdef __GNUC__
+  __attribute__((format(printf, 2, 3)))
+#endif
+;
+
+// Similar to assert(), but uses W_LOG_FATAL to log the stack trace
+// before giving up the ghost
+#ifdef NDEBUG
+# define w_assert(e, ...) ((void)0)
```
Comment (sid0) [inline watchman_log.h]: Ah, guess it's not.

File: pending.c [hunk 8]
```diff
@@ -98,14 +144,132 @@ void w_pending_coll_unlock(struct watchman_pending_collection *coll) {
...
+  unused_parameter(value);
+
+  if ((p->flags & W_PENDING_CRAWL_ONLY) == 0 && key_len > ctx->root->len &&
+      is_path_prefix((const char *)key, key_len, ctx->root->len) &&
+      !is_possibly_a_cookie(p->path)) {
+    // Unlink the child from the pending index.
+    unlink_item(ctx->coll, p);
+    // and completely free it.
+
+#if 0
```
Comment (sid0) [inline pending.c]: Make this a debug call or remove it?

**Comment by sid0** [comment]
`use art to optimize pending list construction` looks fine overall, but shouldn't it be after the artificial-prefix change rather than before? Seems like we're going to run into the prefix issue pretty quickly with the pending list.

File: pending.c [hunk 9]
```diff
@@ -169,21 +339,28 @@ void w_pending_coll_append(struct watchman_pending_collection *target,
...
     }
 
-    p->next = target->pending;
-    target->pending = p;
-    w_ht_set(target->pending_uniq, w_ht_ptr_val(p->path), w_ht_ptr_val(p));
+    if (is_obsoleted_by_containing_dir(target, p->path)) {
+      w_pending_fs_free(p);
+      continue;
+    }
+    prune_obsoleted_children(target, p->path, p->flags);
```
Comment (sid0) [inline pending.c]: One general comment here is that it is probably possible to split this up so that `is_obsoleted_by_containing_dir` is in a separate diff from the `prune_obsoleted_children` function (both really seem to add on to each other). But it's ok :)

File: pending.c [hunk 10]
```diff
@@ -98,14 +144,132 @@ void w_pending_coll_unlock(struct watchman_pending_collection *coll) {
...
+  }
+
+  return 0;
+}
+
+// if there are any entries that are obsoleted by a recursive insert,
+// walk over them now and mark them as ignored.
+static void prune_obsoleted_children(struct watchman_pending_collection *coll,
+                                     w_string_t *path, int flags) {
+  if ((flags & (W_PENDING_RECURSIVE | W_PENDING_CRAWL_ONLY)) ==
```
Comment (sid0) [inline pending.c]: As a casual reader, I'd be surprised that there's a condition in here -- it might make more sense to call this `maybe_prune_obsoleted_children`

File: root.c [hunk 11]
```diff
@@ -498,12 +498,12 @@ bool w_root_process_pending(w_root_t *root,
...
 
   w_log(W_LOG_DBG, "processing %d events in %s\n",
-      w_ht_size(coll->pending_uniq), root->root_path->buf);
+      w_pending_coll_size(coll), root->root_path->buf);
 
   // Steal the contents
   pending = coll->pending;
   coll->pending = NULL;
-  w_ht_free_entries(coll->pending_uniq);
+  w_pending_coll_drain(coll);
```
Comment (sid0) [inline root.c]: I don't understand this change -- could you explain it?

File: tests/pending_test.c [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+    gettimeofday(&start, NULL);
+    for (item = list.avail - 1; item >= list.pending; item--) {
+      w_pending_coll_add(&coll, item->path, item->now, item->flags);
+    }
+
+    drained = process_items(&coll);
+
+    gettimeofday(&end, NULL);
+    diag("took %.3fs to reverse insert %u items into pending coll",
```
Comment (sid0) [inline tests/pending_test.c]: What does "reverse insert" mean here? Might be better to call this "top down" or "bottom up".

File: root.c (hunk 11 above)
Comment (wez) [inline root.c]: This is fixing a mild layering violation.  This bit of code wants to steal the linked list of pending items, but wants to make sure that the pending list doesn't reference them any more.  Previously this was done by ensuring that the pending_uniq hash got emptied, but since that hash was removed by this diff it can't do that any more.

Really this code should have been calling `w_pending_coll_drain`; that function pops anything from the list (which will be empty in this case, because we just stole it) and then empties the hash.

After this diff, we use the `art` instead of the hash.  There isn't an equivalent function to `w_ht_free_entries` for `art`, instead we have to destroy and re-initialize; we could chose to continue the layering violation here, but it would look uglier :-)

File: thirdparty/libart/src/art.c (hunk 3 above)
Comment (wez) [inline thirdparty/libart/src/art.c]: Yeah, not so much at insertion time, but when we perform a prefix search we don't want to have to make a copy, especially if there are no matches for the prefix.

File: thirdparty/libart/src/art.c (hunk 2 above)
Comment (wez) [inline thirdparty/libart/src/art.c]: yeah, it didn't seem like the upstream was very open to making changes in any case, and I've got my eye on a further change to allow holding a reference to our `w_string_t` in the leaves rather than making a copy.

**Comment by facebook-github-bot-0** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@wez updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@wez updated the pull request.
